### PR TITLE
Python: Add Google Cloud Storage support

### PR DESCRIPTION
**Pull request by Buktoria**
Adding Google Cloud Storage support to the Python Iceberg CLI

* Added GCSFileIO support through the `gcsfs` python lib
* Added PyArrow GCS support with PyArrows `GcsFileSystem` class
* Added GCS test suite using a GCS emulator Docker Container (https://github.com/fsouza/fake-gcs-server)
* Updated docs to include GCSFileIO configs options (full list of possible options can be found [here](https://gcsfs.readthedocs.io/en/latest/_modules/gcsfs/core.html#GCSFileSystem))

### CODE REVIEW

**Comment by Fokko** [comment]
@Buktoria this is awesome, thanks for working on this. Let me know when it is ready for review!

**Comment by Buktoria** [comment]
@Fokko This PR is ready to be reviewed

File: python/pyiceberg/io/pyarrow.py [hunk 1]
```diff
@@ -278,6 +279,15 @@ def _get_fs(self, scheme: str) -> FileSystem:
...
             return S3FileSystem(**client_kwargs)
+        elif scheme in {"gs", "gcs"}:
+            if access_token := self.properties.get("gs.access-token"):
+                client_kwargs = {
+                    "access_token": access_token,
+                    "credential_token_expiration": self.properties.get("gs.credential-token-expiration"),
+                }
+                return GcsFileSystem(**client_kwargs)
+            else:
+                return GcsFileSystem()
```
Comment (JonasJ-ap) [inline python/pyiceberg/io/pyarrow.py:290]: [Question] Is it possible to take out this `if...else...`, like:
```python
        elif scheme in {"gs", "gcs"}:
            client_kwargs = {
                "access_token": self.properties.get("gs.access-token"),
                "credential_token_expiration": self.properties.get("gs.credential-token-expiration"),
            }
            return GcsFileSystem(**client_kwargs)
```
?
Is there any problem if `access_token` is not given but `credential_token_expiration` is provided? Thank you in advance for your answer.

**Review by JonasJ-ap — COMMENTED** [review]
Thank you for your great work!

**Comment by Fokko** [comment]
@Buktoria thanks for raising this. I was out for a couple of weeks, could you resolve the merge conflicts? I'll review the PR asap.

File: python/pyiceberg/io/fsspec.py [hunk 2]
```diff
@@ -120,6 +121,24 @@ def _s3(properties: Properties) -> AbstractFileSystem:
...
+        "project": properties.get("gs.project"),
+        "access": properties.get("gs.access", "full_control"),
+        "token": properties.get("gs.token"),
+        "consistency": properties.get("gs.consistency"),
+        "cache_timeout": properties.get("gs.cache-timeout"),
+        "requester_pays": properties.get("gs.requester-pays"),
+        "session_kwargs": json.loads(properties.get("gs.session-kwargs", "{}")),
+        "endpoint_url": properties.get("gs.endpoint-url"),
+        "default_location": properties.get("gs.default-location"),
+        "version_aware": json.loads(properties.get("gs.version-aware", "false").lower()),
```
Comment (Fokko) [inline python/pyiceberg/io/fsspec.py:137]:
```suggestion
        "version_aware": properties.get("gs.version-aware", "false").lower() == 'true',
```

**Review by Fokko — COMMENTED** [review]
@Buktoria Sorry for pinging you again? Would you have any time to fix the merge conflicts? It would be great to get this into 0.4.0

**Comment by Fokko** [comment]
@Buktoria No problem, just checking if it is still on your list. Would be really cool to get this in.

File: python/pyiceberg/io/pyarrow.py [hunk 3]
```diff
@@ -278,6 +279,15 @@ def _get_fs(self, scheme: str) -> FileSystem:
                 "region": self.properties.get(S3_REGION),
             }
             return S3FileSystem(**client_kwargs)
+        elif scheme in {"gs", "gcs"}:
+            if access_token := self.properties.get("gs.access-token"):
```
Comment (danielcweeks) [inline python/pyiceberg/io/pyarrow.py:283]: Can we add `gs.credential-token-expiration` to the docs?

File: python/mkdocs/docs/configuration.md [hunk 4]
```diff
@@ -73,7 +73,22 @@ For the FileIO there are several configuration options available:
...
-| adlfs.client-secret     | oCA3R6P\*ka#oa1Sms2J74z...                                                                 | The client-secret                                                                                                                                                                                                                                                                    |
+| adlfs.client-secret     | oCA3R6P\*ka#oa1Sms2J74z...                                                                | The client-secret                                                                                                                                                                                                                                                                    |
+
+### Google Cloud Storage
```
Comment (bryanck) [inline python/mkdocs/docs/configuration.md:78]: It would be nice if the Python GCS properties could be consistent with the Java properties where possible. Java already has some properties defined so starting with those would help avoid any backwards compatibility issues (rather than changing them). There is also a [PR](https://github.com/apache/iceberg/pull/8073) to add OAuth2 access token properties on the Java side.

File: python/pyiceberg/io/pyarrow.py (hunk 3 above)
Comment (Buktoria) [reply python/pyiceberg/io/pyarrow.py]: I missed it. I've added it to the docs now.

File: python/pyiceberg/io/pyarrow.py (hunk 1 above)
Comment (Buktoria) [reply python/pyiceberg/io/pyarrow.py:314]: If `access_token` is set `credential_token_expiration` also needs to be set.

File: python/mkdocs/docs/configuration.md [hunk 5]
```diff
@@ -74,13 +74,30 @@ For the FileIO there are several configuration options available:
...
 | ----------------------- | ----------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
 | adlfs.connection-string | AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqF...;BlobEndpoint=http://localhost/ | A [connection string](https://learn.microsoft.com/en-us/azure/storage/common/storage-configure-connection-string). This could be used to use FileIO with any adlfs-compatible object storage service that has a different endpoint (like [azurite](https://github.com/azure/azurite)). |
+
```
Comment (Fokko) [inline python/mkdocs/docs/configuration.md:77]: Can you remove this unrelated change?

File: python/mkdocs/docs/index.md [hunk 6]
```diff
@@ -60,7 +60,8 @@ You can mix and match optional dependencies depending on your needs:
 | s3fs     | S3FS as a FileIO implementation to interact with the object store    |
 | adlfs    | ADLFS as a FileIO implementation to interact with the object store   |
 | snappy   | Support for snappy Avro compression                                  |
+| gcs      | GCS as the FileIO implementation to interact with the object store   |
```
Comment (Fokko) [inline python/mkdocs/docs/index.md:63]:
```suggestion
| gcsfs    | GCS as the FileIO implementation to interact with the object store   |
```

**Review by Fokko — COMMENTED** [review]
Thanks for picking this up, it looks like the CI is failing. Can you run `pip3 install pre-commit && pre-commit run --all-files`?

**Comment by Fokko** [comment]
Closing this one, since #8207 is in! Thanks for the PR @Buktoria 🙌🏻
